File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/loader/RawFileItem.java
package ru.kmorozov.gbd.core.loader;

import ru.kmorozov.gbd.core.config.IStoredItem;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Path;

public class RawFileItem implements IStoredItem {

    protected final File outputFile;
    private OutputStream outputStream;

    public RawFileItem(File file) {
        this.outputFile = file;
    }

    public RawFileItem(Path path) {
        this(path.toFile());
    }

    @Override
    public OutputStream getOutputStream() throws IOException {
        return outputStream = outputStream == null ? new FileOutputStream(outputFile) : outputStream;
    }

    @Override
    public boolean exists() throws IOException {
        return outputFile.exists();
    }

    @Override
    public void delete() throws IOException {
        outputFile.delete();
    }

    @Override
    public void close() throws IOException {
        if (outputStream != null)
            outputStream.close();
    }

    @Override
    public void write(byte[] bytes, int read) throws IOException {
        getOutputStream().write(bytes, 0, read);
    }

    @Override
    public File asFile() {
        return outputFile;
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/loader/RawFileItem.kt
package ru.kmorozov.gbd.core.loader

import ru.kmorozov.gbd.core.config.IStoredItem
import ru.kmorozov.gbd.utils.Images
import java.io.*
import java.nio.file.Path

open class RawFileItem : IStoredItem {
    protected val outputFile: File

    override val createdNew: Boolean

    constructor(outputFile: File) {
        this.outputFile = outputFile
        this.createdNew = !outputFile.exists()
    }

    constructor(path: Path) : this(path.toFile()) {}

    override lateinit var outputStream: OutputStream

    var totalLen = 0L

    override val pageNum: Int
        get() = Integer.parseInt(outputFile.toPath().fileName.toString().split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0])

    @Throws(IOException::class)
    override fun exists(): Boolean {
        return outputFile.exists()
    }

    @Throws(IOException::class)
    override fun delete() {
        try {
            if (totalLen > 0) {
                totalLen = 0;
            }

            outputFile.delete()
        } catch (ex: IOException) {
            ex.printStackTrace()
        }
    }

    @Throws(IOException::class)
    override fun close() {
    }

    @Throws(IOException::class)
    override fun write(inStream: InputStream) {
        FileOutputStream(outputFile).use { totalLen += inStream.transferTo(it) }
    }

    open fun init() {

    }

    override fun flush() {

    }

    override fun asFile(): File {
        return outputFile
    }

    override fun isImage(): Boolean {
        return Images.isImageFile(outputFile.toPath())
    }

    override fun validate(): Boolean {
        return outputFile.exists()
    }
}
